master/internal/rm/kubernetesrm/pods.go

```diff
@@ -864,6 +908,157 @@ func (p *pods) podStatusCallback(s *actor.System, event watch.Event) {
+			"labels": {
+				"%s": null
+			}
+		}
+	}`, clusterIDNodeLabel()))
+
+	_, err := p.clientSet.CoreV1().Nodes().
+		Patch(context.TODO(), nodeName, types.StrategicMergePatchType, patch, metaV1.PatchOptions{})
+	if k8error.IsForbidden(err) {
+		return nil, fmt.Errorf("the Determined master Kubernetes service account " +
+			"is missing permissions to patch nodes. " +
+			"Enabling or disabling nodes requires this permission, " +
+			"however Determined will otherwise still function correctly without " +
+			"these Kubernetes permissions")
+	} else if err != nil {
+		return nil, fmt.Errorf(
+			"enabling node %s by removing the Determined no schedule label: %w", nodeName, err)
+	}
+	p.syslog.Infof("node %s enabled by an user", nodeName)
+
+	n, ok := p.summarizeClusterByNodes(ctx)[nodeName]
+	if !ok {
+		return nil, fmt.Errorf("node %s enabled without error, error getting node summary", nodeName)
+	}
+	n.Enabled = true
+	n.Draining = false
+	for slotKey := range n.Slots {
+		s := n.Slots[slotKey]
+		s.Enabled = n.Enabled
+		s.Draining = n.Draining
+		n.Slots[slotKey] = s
+	}
+
+	return &apiv1.EnableAgentResponse{
+		Agent: n.ToProto(),
+	}, nil
+}
+
+func (p *pods) disableNode(
+	ctx *actor.Context, nodeName string, shouldDrain bool,
+) (*apiv1.DisableAgentResponse, error) {
+	labelValue := noExecuteNodeLabelValue
+	if shouldDrain {
+		labelValue = noScheduleNodeLabelValue
+	}
+
+	patchStruct := metaV1.ObjectMeta{
+		Labels: map[string]string{clusterIDNodeLabel(): labelValue},
+	}
+	patch, err := json.Marshal(map[string]any{"metadata": patchStruct})
+	if err != nil {
+		return nil, fmt.Errorf("marshaling JSON patch %v: %s", patchStruct, err)
+	}
+
+	_, err = p.clientSet.CoreV1().Nodes().
+		Patch(context.TODO(), nodeName, types.StrategicMergePatchType, patch, metaV1.PatchOptions{})
+	if k8error.IsForbidden(err) {
+		return nil, fmt.Errorf("the Determined master Kubernetes service account " +
+			"is missing permissions to patch nodes. " +
+			"Enabling or disabling nodes requires this permission, " +
+			"however Determined will otherwise still function correctly without " +
+			"these Kubernetes permissions")
+	} else if err != nil {
+		return nil, fmt.Errorf(
+			"disabling node %s by adding the Determined no schedule label: %w", nodeName, err)
+	}
+	p.syslog.Infof("node %s disabled by an user", nodeName)
+
+	if !shouldDrain { // See note in spec.go about how we could remove killing all pods here.
+		if err := p.releaseAllocationsOnDisabledNode(ctx, nodeName); err != nil {
+			return nil, fmt.Errorf(
+				"node disabled without error, error killing existing pod on node: %w", err)
+		}
+	}
+
+	n, ok := p.summarizeClusterByNodes(ctx)[nodeName]
+	if !ok {
+		return nil, fmt.Errorf("node %s disabled without error, error getting node summary", nodeName)
+	}
+	n.Enabled = false
+	n.Draining = shouldDrain
+	for slotKey := range n.Slots {
+		s := n.Slots[slotKey]
+		s.Enabled = n.Enabled
+		s.Draining = n.Draining
+		n.Slots[slotKey] = s
+	}
```

Comment:
if you are going to mutate the contents of the node cache, use a lock. or better yet, make a copy and reset it.

the original intent was that the cache was immutable.

Reply:
https://github.com/determined-ai/determined/blob/84a6612fe68264a0fb8040f46dfc9500a41510d7/master/internal/rm/kubernetesrm/pods.go#L1180C16-L1180C39

confusingly, but also intentionally this doesn't use the node cache I think 

I originally did have the node cache reset the timestamp then query it, but I removed that since in part I think there is no guarantee that our node informer picks up the update by the time we query it?

Maybe I'll add back in the invalidation after just since someone is likely to call ```det a``` after ```det enable/disable```

Reply:
The intention of this n.Enabled code is just so we don't return wrong results in the proto def

This is proto stuff though that kinda is annoying, since we don't use this result anywhere but still don't want to feel like I'm returning lies in case some person decides to depend on it sometime

Reply:
Oh now I remember, GetAgents itself doesn't use the cache

Reply:
chatted offline, this is a bug. we'll file a ticket and fix it soon.

Reply:
https://hpe-aiatscale.atlassian.net/browse/DET-9836